### ISSUE TITLE
Make a backup of module configuration

### ISSUE THREAD

**Issue by SvenRosvall** [david284/MMC-SERVER#23]
This need comes out of a MERG forum thread: https://www.merg.org.uk/forum/viewtopic.php?p=213423
Here they discuss how to replace a faulty node on a layout with a fresh new node and configuring that new node with the NVs and events needed to do the same job.

MMC doesn't store the full node configuration on disk. The main idea is to use the config on the nodes to avoid duplication of the configuration that can get out of sync. But there is a need to be able to replace nodes on a layout bus. Taking a backup of the configuration is one way. Of course MMC also needs to be able to write that backed up configuration to a new node.

This feature requires a lot of work and is not required for normal operation. So it may have low priority. But it will be a required feature to be able to replace FCU.

**Comment by david284** [comment]
I believe the current operation of the FCU stores the 'last' read configuration of nodes, so a 'replacement' node can be written with this 'last' read config, as long as another read hasn't been performed - this works as the FCU has to be manually triggered to read the nodes, unlike the MMC, which does this automatically
However, the FCU doesn't have the capability to 'revert' to a previous configuration - it doesn't have any storage of 'older' configurations, which I believe would be a useful feature

So, I would propose storing more than one configuration file - in a similar manor to 'restore points' in windows
I assume each individual configuration would be identified by date&time, node number and then module Id
(as there could be multiple nodes of the same type, and each config would be specific to module type, so there'd need to be a check for suitability)

There will need to be a choice made on when to store these files - automatically on some trigger, or manual selection
(automatically on every config change could generate an excessive number of files) - I believe the FCU prompts on shutdown - this doesn't sound unreasonable
Obviously a 'restore previous config' option would need to be added, but I suspect that will be heavily influenced by the storage method

**Comment by SvenRosvall** [comment]
Yes, this is true about FCU. That means in effect that FCU doesn't have a proper backup system, i.e. ability to roll back to some old state. FCU can be used to replace a node but not roll back a bad configuration.

Agree with "restore points" and keeping many backups. This feature request needs to be pondered over, brainstormed, and discussed. And again, this is low priority.

Duncan mentioned at the last VLCB meeting that JMRI saves backups automatically. He compared this to how Windows creates recovery points.

**Comment by SvenRosvall** [comment]
Had a converstationg with Greg Palmer. He suggested that a backup or some other means of captured configurations would be useful for:
- Reporting bugs / sharing configuration with others.
- Get an overview of the configs. (UI doesn't (and cannot) show all configs in one screen.)
- Validation. A tool reads the backup/config file. This is already done with the config file from FCU.

**Comment by david284** [comment]
done
a5058cfb9cfc2c15b743ad69cb160482dbe8c324
